src/Services.php

```diff
@@ -50,34 +52,34 @@ public static function registerServices(ContainerBuilder $container) {
+    $container->setDefinition('civi_flexmailer_default_batcher', new Definition('Civi\FlexMailer\Listener\DefaultBatcher'))->setPublic(TRUE);
+    $container->setDefinition('civi_flexmailer_default_composer', new Definition('Civi\FlexMailer\Listener\DefaultComposer'))->setPublic(TRUE);
+    $container->setDefinition('civi_flexmailer_open_tracker', new Definition('Civi\FlexMailer\Listener\OpenTracker'))->setPublic(TRUE);
+    $container->setDefinition('civi_flexmailer_basic_headers', new Definition('Civi\FlexMailer\Listener\BasicHeaders'))->setPublic(TRUE);
+    $container->setDefinition('civi_flexmailer_to_header', new Definition('Civi\FlexMailer\Listener\ToHeader'))->setPublic(TRUE);
+    $container->setDefinition('civi_flexmailer_attachments', new Definition('Civi\FlexMailer\Listener\Attachments'))->setPublic(TRUE);
+    $container->setDefinition('civi_flexmailer_bounce_tracker', new Definition('Civi\FlexMailer\Listener\BounceTracker'))->setPublic(TRUE);
+    $container->setDefinition('civi_flexmailer_default_sender', new Definition('Civi\FlexMailer\Listener\DefaultSender'))->setPublic(TRUE);
+    $container->setDefinition('civi_flexmailer_hooks', new Definition('Civi\FlexMailer\Listener\HookAdapter'))->setPublic(TRUE);
+    $container->setDefinition('civi_flexmailer_test_prefix', new Definition('Civi\FlexMailer\Listener\TestPrefix'))->setPublic(TRUE);
+
+    $container->setDefinition('civi_flexmailer_html_click_tracker', new Definition('Civi\FlexMailer\ClickTracker\HtmlClickTracker'))->setPublic(TRUE);
+    $container->setDefinition('civi_flexmailer_text_click_tracker', new Definition('Civi\FlexMailer\ClickTracker\TextClickTracker'))->setPublic(TRUE);
 
     foreach (self::getListenerSpecs() as $listenerSpec) {
-      $container->findDefinition('dispatcher')->addMethodCall('addListenerService', $listenerSpec);
+      $container->findDefinition('dispatcher')->addMethodCall('addListener', [$listenerSpec[0], [new ServiceClosureArgument(new Reference($listenerSpec[1][0])), $listenerSpec[1][1]], $listenerSpec[2]]);
     }
 
     $container->findDefinition('civi_api_kernel')->addMethodCall('registerApiProvider', array(new Reference('civi_flexmailer_api_overrides')));
   }
 
-  public static function registerListeners(ContainerAwareEventDispatcher $dispatcher) {
+  public static function registerListeners(EventDispatcher $dispatcher) {
     foreach (self::getListenerSpecs() as $listenerSpec) {
-      $dispatcher->addListenerService($listenerSpec[0], $listenerSpec[1], $listenerSpec[2]);
-      $dispatcher->addSubscriberService('civi_flexmailer_api_overrides', 'Civi\FlexMailer\API\Overrides');
+      $dispatcher->addListener($listenerSpec[0], [new SercviceClosureArgument(new Reference($listenerSpec[1][0])), $listenerSpec[1][1]], $listenerSpec[2]);
+      $dispatcher->addSubscriber(new Reference('civi_flexmailer_api_overrides'));
```

Comment:
Aah, so this example change is much clearer than the upstream docs. :+1: I'm not certain that these lines work per se (more in a sec), but it was enlightening to read.

Digging in on this PR/function was a bit of rabbit hole...

* Neither the [upstream docs](https://symfony.com/doc/4.4/components/event_dispatcher/container_aware_dispatcher.html], [changelog](https://github.com/symfony/symfony/blob/3.3/CHANGELOG-3.3.md), [deprecation PR](https://github.com/symfony/symfony/pull/20937), nor [docs issue](https://github.com/symfony/symfony/issues/22402) provide an example snippet or reference to the alternative helpers. Kudos to you for figuring a work-a-like in Symfony 4.x's class model.
* There's a typo in the class name (`Sercvice`). Digging into this revealed that the function `registerListeners()` is never actually called. It's leftover from some alpha/beta hackery and was disabled (17f142e1539ac2e507d35b837607b8fad965b6a5) when `hook_civicrm_container` became more reliable. This also explains things worked despite some more flaws in the old variant of `registerListeners()`.
* How does `civi_flexmailer_api_overrides` work *at all*? Well, it does. ✅  It's because there's a call to `registerApiProvider()`, and *that* is what really handles registration.